Profiler as lifecycle hook

gregmagolan opened the pull request:
@achew22 

Let me know what you think.

I renamed `TargetDecider` to `RunTarget` as I think the former is confusing given that it is only called for `run` targets and not for `build` and `test` targets. Personally don't think there is that much value in having iBazel keep track of which lifecycle listeners are turned on/off. Seems like an anti-pattern. Its cleaner to have the listeners themselves worry about which events they care about and which ones to ignore.

IBazel itself conforms to the Lifecycle interface so it can be passed in to the certain lifecycle hooks (ChangeDetected, BeforeCommand, AfterCommand and ReloadTriggered right now) so that lifecycle events can create other lifecycle events. This is use inside livereload AfterCommand which fires off a ReloadTriggered event that is consumed by the profiler.

Also added a few some more parameters to various events that listeners can choose to use.

Usage looks like this: `ibazel -profile_dev=profile.out run devserver`

Output file looks something like the following in concatenated JSON with the `iteration` UUID marking a command loop in iBazel. This output includes REMOTE_EVENTs that come back from the browser:

```
{"type":"IBAZEL_START","iteration":"b595909a-33d7-4bbe-a5a0-808d0d187bd7","time":1513021938425,"bazelVersion":"release 0.8.1-homebrew","maxHeapSize":"3817MB","committedHeapSize":"1141MB"}
{"type":"RUN_START","iteration":"b595909a-33d7-4bbe-a5a0-808d0d187bd7","time":1513021938825,"targets":["devserver"],"elapsed":400}
{"type":"RELOAD_TRIGGERED","iteration":"b595909a-33d7-4bbe-a5a0-808d0d187bd7","time":1513021939462,"targets":["devserver"],"elapsed":1037}
{"type":"RUN_DONE","iteration":"b595909a-33d7-4bbe-a5a0-808d0d187bd7","time":1513021939462,"targets":["devserver"],"elapsed":1037}
{"type":"REMOTE_EVENT","iteration":"b595909a-33d7-4bbe-a5a0-808d0d187bd7","time":1513021942099,"targets":["devserver"],"elapsed":3674,"remoteType":"PAGE_LOAD","remoteTime":1513021942095,"remoteElapsed":379,"remoteData":"{\"pageLoadTime\":379,\"fetchTime\":11,\"connectTime\":0,\"requestTime\":9,\"responseTime\":4,\"renderTime\":357,\"navigationStart\":1513021941716,\"unloadEventStart\":1513021941730,\"unloadEventEnd\":1513021941732,\"redirectStart\":0,\"redirectEnd\":0,\"fetchStart\":1513021941719,\"domainLookupStart\":1513021941721,\"domainLookupEnd\":1513021941721,\"connectStart\":1513021941721,\"connectEnd\":1513021941721,\"secureConnectionStart\":0,\"requestStart\":1513021941721,\"responseStart\":1513021941726,\"responseEnd\":1513021941730,\"domLoading\":1513021941738,\"domInteractive\":1513021942025,\"domContentLoadedEventStart\":1513021942025,\"domContentLoadedEventEnd\":1513021942026,\"domComplete\":1513021942095,\"loadEventStart\":1513021942095}"}

{"type":"SOURCE_CHANGE","iteration":"2e8370f1-2d81-4950-8728-70a1fa804de2","time":1513021966893,"targets":["devserver"],"change":"/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts"}
{"type":"RUN_START","iteration":"2e8370f1-2d81-4950-8728-70a1fa804de2","time":1513021966995,"targets":["devserver"],"elapsed":102,"changes":["/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts"]}
{"type":"RELOAD_TRIGGERED","iteration":"2e8370f1-2d81-4950-8728-70a1fa804de2","time":1513021970618,"targets":["devserver"],"elapsed":3725,"changes":["/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts"]}
{"type":"RUN_DONE","iteration":"2e8370f1-2d81-4950-8728-70a1fa804de2","time":1513021970618,"targets":["devserver"],"elapsed":3725,"changes":["/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts"]}
{"type":"REMOTE_EVENT","iteration":"2e8370f1-2d81-4950-8728-70a1fa804de2","time":1513021971005,"targets":["devserver"],"elapsed":4112,"remoteType":"PAGE_LOAD","remoteTime":1513021971001,"remoteElapsed":382,"remoteData":"{\"pageLoadTime\":381,\"fetchTime\":7,\"connectTime\":0,\"requestTime\":5,\"responseTime\":5,\"renderTime\":363,\"navigationStart\":1513021970619,\"unloadEventStart\":1513021970630,\"unloadEventEnd\":1513021970631,\"redirectStart\":0,\"redirectEnd\":0,\"fetchStart\":1513021970622,\"domainLookupStart\":1513021970622,\"domainLookupEnd\":1513021970622,\"connectStart\":1513021970622,\"connectEnd\":1513021970622,\"secureConnectionStart\":0,\"requestStart\":1513021970624,\"responseStart\":1513021970624,\"responseEnd\":1513021970629,\"domLoading\":1513021970637,\"domInteractive\":1513021970943,\"domContentLoadedEventStart\":1513021970943,\"domContentLoadedEventEnd\":1513021970944,\"domComplete\":1513021971000,\"loadEventStart\":1513021971000}"}

{"type":"SOURCE_CHANGE","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021986328,"targets":["devserver"],"change":"/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts"}
{"type":"SOURCE_CHANGE","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021986328,"targets":["devserver"],"change":"/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer1.ts"}
{"type":"SOURCE_CHANGE","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021986328,"targets":["devserver"],"change":"/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer2.ts"}
{"type":"RUN_START","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021986429,"targets":["devserver"],"elapsed":101,"changes":["/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts","/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer1.ts","/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer2.ts"]}
{"type":"RELOAD_TRIGGERED","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021988091,"targets":["devserver"],"elapsed":1763,"changes":["/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts","/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer1.ts","/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer2.ts"]}
{"type":"RUN_DONE","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021988091,"targets":["devserver"],"elapsed":1763,"changes":["/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer0.ts","/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer1.ts","/Users/greg/google/gregmagolan/webpack-typescript-benchmarking/src/package0/timer2.ts"]}
{"type":"REMOTE_EVENT","iteration":"dc9c57a7-33f1-46ad-8827-fe07ccfd348f","time":1513021988469,"targets":["devserver"],"elapsed":2141,"remoteType":"PAGE_LOAD","remoteTime":1513021988465,"remoteElapsed":373,"remoteData":"{\"pageLoadTime\":373,\"fetchTime\":10,\"connectTime\":0,\"requestTime\":8,\"responseTime\":7,\"renderTime\":348,\"navigationStart\":1513021988092,\"unloadEventStart\":1513021988105,\"unloadEventEnd\":1513021988106,\"redirectStart\":0,\"redirectEnd\":0,\"fetchStart\":1513021988095,\"domainLookupStart\":1513021988095,\"domainLookupEnd\":1513021988095,\"connectStart\":1513021988095,\"connectEnd\":1513021988095,\"secureConnectionStart\":0,\"requestStart\":1513021988097,\"responseStart\":1513021988098,\"responseEnd\":1513021988105,\"domLoading\":1513021988117,\"domInteractive\":1513021988405,\"domContentLoadedEventStart\":1513021988405,\"domContentLoadedEventEnd\":1513021988406,\"domComplete\":1513021988465,\"loadEventStart\":1513021988465}"}
```

bazel-io commented:
Can one of the admins verify this patch?

gregmagolan commented:
Replaces #87

gregmagolan commented:
@alexeagle Profiler client-side script is served from /profiler.js. To turn this on in devserver it will be a small change:

```
	// Include the profiler script if IBAZEL_PROFILER_URL is set.
	profilerUrl := os.Getenv("IBAZEL_PROFILER_URL")
	if profilerUrl != "" {
		fmt.Printf("Serving profiler script from %s\n", profilerUrl)
		profilerLoaderSnippet := fmt.Sprintf(`(function(){
	const script = document.createElement('script');
	script.src = "%s";
	document.head.appendChild(script);
})();`, profilerUrl)
		preScripts = append(preScripts, profilerLoaderSnippet)
	}
```

gregmagolan commented on WORKSPACE:
```diff
@@ -77,3 +77,9 @@ go_repository(
     importpath = "github.com/gorilla/websocket",
 )
+
+go_repository(
```
No reason. It could be a random hex string. Very little chance of collision. Will change that.

gregmagolan commented on ibazel/live_reload/server.go:
```diff
@@ -82,25 +59,47 @@ func (l *LiveReloadServer) TargetDecider(rule *blaze_query.Rule) {
 	if l.lrserver != nil {
 		fmt.Fprintf(os.Stderr, "Triggering live reload\n")
 		l.lrserver.Reload("reload")
+		c.ReloadTriggered(c, targets)
```
That's the tricky bit. Unless the the profiler is aware of whether or not livereload is enabled (`if l.lrserver != nil` above) then it can't know if a livereload was triggered. Hence the modifications to allow for lifecycle events to create other lifecycle events by having them take a Lifecycle interface that can be calle.

gregmagolan replied:
I'm open to other suggestions. Presently the profiler is using the livereload event to know if it should accept REMOTE_EVENT from the browser. If live reload is disabled then it ignore the REMOTE_EVENTs as they would be user driven and not useful to profiling the RTT.

gregmagolan commented on ibazel/profiler/javascript.go:
```diff
@@ -0,0 +1,126 @@
+package profiler
+
+const js string = `(function(){
+  function setOption(options, name, value) {
```
New code. The client-side script that generates the REMOTE_EVENTs that the profiler listens for.

gregmagolan commented on ibazel/profiler/profiler.go:
```diff
@@ -0,0 +1,396 @@
+
+type Profiler struct {
+	server                   *http.Server
+	f                        *os.File
```
Sounds good

gregmagolan commented on ibazel/profiler/profiler.go:
```diff
@@ -0,0 +1,396 @@
+
+	ln.Close()
+	return true
+}
```
Yup. That was the next step but first wanted to check if I was going in the right direction with this.

gregmagolan commented on ibazel/ibazel.go:
```diff
@@ -26,7 +26,9 @@ import (
 
 	"github.com/bazelbuild/bazel-watcher/bazel"
 	"github.com/bazelbuild/bazel-watcher/ibazel/command"
+	"github.com/bazelbuild/bazel-watcher/ibazel/lifecycle"
```
This was needed to avoid circular deps. Since I'm passing the Lifecycle interface into some of the hooks, the lifecycle listeners needed to have access to the interface.

gregmagolan commented on ibazel/ibazel.go:
```diff
@@ -248,17 +256,22 @@ func (i *IBazel) iteration(command string, commandToRun runnableCommand, targets
 		case e := <-i.sourceEventHandler.SourceFileEvents:
 			if e.Op&modifyingEvents != 0 {
 				fmt.Fprintf(os.Stderr, "Changed: %q. Rebuilding...\n", e.Name)
+				i.ChangeDetected(i, targets, "source", e.Name)
```
That could work. A different interface that contains only side-effect lifecycle hooks like ReloadTriggered.

gregmagolan commented on ibazel/profiler/profiler.go:
```diff
@@ -0,0 +1,396 @@
+}
+
+func (i *Profiler) reloadTriggeredEvent() {
+	i.lock.Lock()
```
There a few variables that need protecting as the profiler is accepting remote events via http. Basically anything that gets touched in `remoteEvent` and `processEvent`: ```func (i *Profiler) remoteEvent(remoteEvent *profilerRemoteEvent) {
	i.lock.Lock()
	if !i.iterationReloadTriggered {
		fmt.Fprintf(os.Stderr, "Ignoring unexpected remote event\n")
		return
	}
	event := profileEvent{}
	event.Type = "REMOTE_EVENT"
	event.RemoteType = remoteEvent.Type
	event.RemoteTime = remoteEvent.Time
	event.RemoteElapsed = remoteEvent.Elapsed
	event.RemoteData = remoteEvent.Data
	i.processEvent(&event)
	i.lock.Unlock()
}
```

`i.iterationReloadTriggered`, `i.iteration`, `i.targets`, `i.iterationStartTime`

gregmagolan commented on ibazel/profiler/javascript.go:
```diff
@@ -0,0 +1,126 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
```
I like that

gregmagolan commented on ibazel/profiler/profiler.go:
```diff
@@ -0,0 +1,396 @@
+		}
+
+		rw.Header().Set("Content-Type", "application/javascript")
+		_, err := rw.Write([]byte(js))
```
👍

gregmagolan commented on ibazel/ibazel.go:
```diff
@@ -161,31 +167,33 @@ func (i *IBazel) Cleanup() {
 	}
 }
 
-func (i *IBazel) targetDecider(rule *blaze_query.Rule) {
```
I commented on this in the PR comment. `targetDecider` may not be the best name since its only called right now for `run` targets and not for `build` or `test`. Hence the rename so its clear for listeners when to expect that hook. Maybe `runTargetDecider` is better?

gregmagolan commented on ibazel/ibazel.go:
```diff
@@ -161,31 +167,33 @@ func (i *IBazel) Cleanup() {
+func (i *IBazel) ReloadTriggered(c lifecycle.Lifecycle, targets []string) {
 	for _, l := range i.lifecycleListeners {
-		l.AfterCommand(name, success)
+		l.ReloadTriggered(c, targets)
```
ReloadTriggered is meant to only called when the livereload server is enabled and a "reload" event has been sent to the browser.

gregmagolan commented:
@achew22 Update the PR so that the profiler listens for an the ReloadTriggered event from livereload itself instead of going throught the Lifecycle interface. This will be a much cleaner pattern. Stay tuned.

gregmagolan commented:
@achew22 Ok. Its been updated with a better pattern. Ready for another look when you have a chance. I'll work on an e2e in the meantime.

gregmagolan commented:
@achew22 Ping. Can you have another look when you have time? Would like to land this next week as part of the current ABC sprint.

gregmagolan commented:
Profiler e2e test added. Also updated the livereload e2e test to check the /livereload.js result instead of ibazel_test.go doing it.

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+    _ref = window.document.getElementsByTagName('script');
+    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
+      element = _ref[_i];
+      if ((src = element.src) && (m = src.match(/^[^:]+:\/\/(.*)\/z?profiler\.js(?:\?(.*))?$/))) {
```
Since you're in the context of the browser here, I would prefer if you used the browser's url parser to achieve this. I believe this is equivalent to the for loop contents.

```
const parsed = document.createElement('a');
parsed.href = element.src;
options.url = `${parsed.protocol}//${parsed.host}/profiler-event`;
// Note starts here
p.search.substring(1).split('&').forEach((e) => {
  [k, ...v] = e.split("=");
  options[k] = v.join("=");
})
```

Additionally after "// Note starts here" I'm not sure what the value of adding the query params is? Does that exist so that you can override options.{https,host,port}? If you don't care about that feel free to omit after `// Note starts here`

gregmagolan replied:
Thanks.

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  window.IBazelProfileEvent = profilerEvent;
```
This line is unnecessary (unless it's part of a public API being used elsewhere) since on line 91 it should be able to refer to that through the lexical scope of the parent closure.

gregmagolan replied:
Yes. Its meant to be a public API so the user space can call `window.IBazelProfileEvent("MY_EVENT_TYPE", myData)` and that would end up in the profile log as a `REMOTE_EVENT` of `remoteType` `MY_EVENT_TYPE`

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+          }
+        }
+        options.url = (options.https ? "https" : "http") + "://" + options.host + ":" + options.port + "/profiler-event";
+        return options;
```
It appears the only value ever read out of options is the url value. I think you should just return the url or an empty string if it can't find one.

gregmagolan replied:
Yup. Only the url is needed. Cleaned it up.

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+    return null;
+  };
+
+  const profilerOptions = extractOptions();
```
const profilerEventUrl = extractUrl()

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+    return;
+  }
+
+  if (!navigator.sendBeacon) {
```
It seems navigator is in the window scope. Can you explicitly say window.navigator.sendBeacon?

gregmagolan replied:
Sound good

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+  function extractOptions() {
+    var element, keyAndValue, m, mm, options, pair, src, _i, _j, _len, _len1, _ref, _ref1;
+    _ref = window.document.getElementsByTagName('script');
+    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
```
`.getElementsByTagName` returns a live [HTMLCollection](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection) which means you can't directly iterate over it with a for of, but you can do

```
[...window.document.getElementsByTagName('script')].forEach(...)
```

which does some of the bookkeeping for you and is a bit cleaner IMO

gregmagolan replied:
Thanks. That's nice and clean.

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+  }
+
+  function profilerEvent(eventType, eventData) {
+    if (!eventType || typeof eventType !== 'string') {
```
typeof always returns a string, `typeof eventType != 'string'`

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+      console.error("Invalid iBazel profiler event type", eventType);
+      return;
+    }
+    if (typeof eventData === 'object') {
```
Same here

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+      eventData = JSON.stringify(eventData)
+    }
+    const event = { type: eventType, time: Date.now() };
+    event.elapsed = event.time - window.performance.timing.navigationStart;
```
this isn't really the elapsed time, it is the time at which the event occurred in ms relative to page load. Elapsed implies that you were monitoring the event from start to finish.

gregmagolan replied:
Its elapsed time since the navigationStart event so it would be the remote elapsed time I suppose. Renamed to timeSinceNavigationStart but its a bit verbose. Have a better idea? In iBazel, this can value gets passed into remoteElapsedTime in the REMOTE_EVENT

gregmagolan commented on ibazel/lifecycle.go:
```diff
@@ -9,24 +9,24 @@ import (
 	// BeforeCommand is called before a blaze $COMMAND is run.
 	// command: "build"|"test"|"run"
-	BeforeCommand(command string)
+	BeforeCommand(targets []string, command string)
```
Using targets in the profiler as part of the profiler output

gregmagolan commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+    if (typeof value === 'undefined') {
+      return;
+    }
+    if (!isNaN(+value)) {
```
I cleaned up the javascript and removed the unnecessary code including code that would parse the query params as they are not used right now anyway

gregmagolan commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,123 @@
+  };
+
+  function extractOptions() {
+    var element, keyAndValue, m, mm, options, pair, src, _i, _j, _len, _len1, _ref, _ref1;
```
👍

gregmagolan commented on ibazel/live_reload/server.go:
```diff
@@ -29,46 +29,29 @@ import (
 func New() *LiveReloadServer {
-	return &LiveReloadServer{}
+	l := &LiveReloadServer{}
+	l.eventListeners = make([]Events, 0, 10)
```
👍

gregmagolan commented on e2e/profiler/profiler_test.go:
```diff
@@ -0,0 +1,162 @@
+		t.Error(err)
+	}
+
+	client := new(http.Client)
```
👍

gregmagolan commented on e2e/live_reload/live_reload_test.go:
```diff
@@ -81,11 +84,39 @@ sh_binary(
 		t.Error(err)
 	}
 
+	client := new(http.Client)
```
👍

gregmagolan commented on e2e/ibazel.go:
```diff
@@ -81,3 +69,27 @@ func (i *IBazelTester) Kill() {
 }
+
+func (i *IBazelTester) run(target string, profiler string) {
+	if profiler != "" {
```
Thanks. Ended up being 
```
	args := []string{"--bazel_path="+i.bazelPath(), "--log_to_file=/tmp/ibazel_output.log"}
	if profiler != "" {
		args = append(args, "--profile_dev="+profiler)
	}
	args = append(args, "run")
	args = append(args, target)
	i.cmd = exec.Command(ibazelPath, args...)
```
as the `--profile_dev=...` needs to be before `run target`

gregmagolan commented on e2e/profiler/profiler_test.go:
```diff
@@ -0,0 +1,162 @@
+)
+`))
+
+	os.Remove("/tmp/ibazel_output.json")
```
👍

gregmagolan commented:
@achew22 Thanks for you comments. Cleaned up the code and I think I addressed everything you mentioned.

Renamed `event.elapsed` to `event.timeSinceNavigationStart` but it seems a bit verbose so maybe you have a better name?

gregmagolan commented:
I'll update the README.md today with some details of how to use the profile_dev mode. Seems like the best place for the documentation on it.

gregmagolan commented:
@achew22 readme updated with --profile_dev documentation

achew22 commented on e2e/profiler/profiler_test.go:
```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		t.Errorf("Failed to decode RUN_DONE event: %v", err)
+	}
+	assertEqual(t, event.Type, "RUN_DONE", "Expected RUN_DONE")
```
This is a great test!

achew22 commented on ibazel/live_reload/server.go:
```diff
@@ -82,22 +65,50 @@ func (l *LiveReloadServer) TargetDecider(rule *blaze_query.Rule) {
 	if l.lrserver != nil {
 		fmt.Fprintf(os.Stderr, "Triggering live reload\n")
 		l.lrserver.Reload("reload")
+		l.reloadTriggered(targets)
```
I think I would move `l.lrserver.Reload("reload")` into `l.reloadTriggered` to ensure you we never accidentally call reload without notifying the profiler

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+(function(){
+  function extractUrl() {
+    var url;
+    [...window.document.getElementsByTagName('script')].forEach(function (element) {
```
Can you run this through clang format? I expect this line to be:

```
[...window.element..........]
    .forEach(function(element) {
  if (!url && element.src
```

and a similar rendering on line 19. I defer to clang-format so if it disagrees with me then ship that.

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+  function extractUrl() {
+    var url;
+    [...window.document.getElementsByTagName('script')].forEach(function (element) {
+      if (!url && element.src && element.src.match(/^[^:]+:\/\/(.*)\/z?profiler\.js(?:\?(.*))?$/)) {
```
I think you can avoid the regexp and get a slight speedup by doing

```
const parsed = document.createElement('a');
if (parsed.pathname == "/profiler.js") {
  return;
}
...
```

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+        url = `${parsed.protocol}//${parsed.host}/profiler-event`;
+      }
+    });
+    return url;
```
I think the return value to functions passed into `Array.prototype.forEach` is never used so you can save these bytes.

gregmagolan replied:
This is the return value of `function extractUrl()`

gregmagolan replied:
I'll do it the stylistic way with map, filter and reduce instead. Good suggestion

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  if (!window.navigator.sendBeacon) {
+    console.error("iBazel profiler disabled because beacon API is not available")
```
Can you add a link to the README.md and some words in README about what a beacon is so that if you get this error you have some next steps?

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+
+(function(){
+  function extractUrl() {
+    var url;
```
Prefer `let` over `var`. I didn't comment on this before since I thought you might be targeting ES5 browsers, but since you have `const` above you're not going to work with them and `let` is less of a footgun.

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+(function(){
+  function extractUrl() {
+    var url;
+    [...window.document.getElementsByTagName('script')].forEach(function (element) {
```
If you're interested in a stylistic flourish and wanna play around, you could change the `.forEach` into a `.map`, `.filter` all out the urls that are an empty string, and use a `.reduce` to coalesce the result into a single value.

You can totally ignore this one.

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+    }
+    const data = JSON.stringify(event)
+    if (!window.navigator.sendBeacon(profilerEventUrl, data)) {
+      console.error("Failed to send profile data to " + profilerEventUrl);
```
Can you use a , instead of a + so the profilerEventUrl is pretty printed on the console?

gregmagolan replied:
👍

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,97 @@
+  window.IBazelProfileEvent = profilerEvent;
+
+  window.addEventListener("load", function() {
+    var timing = window.performance.timing;
```
var => let

gregmagolan replied:
👍

gregmagolan commented on e2e/ibazel.go:
```diff
@@ -117,3 +105,29 @@ func (i *IBazelTester) Kill() {
 	}
 }
+
+func (i *IBazelTester) run(target string, profiler string) {
```
👍

gregmagolan commented on e2e/live_reload/live_reload_test.go:
```diff
@@ -77,11 +80,39 @@ sh_binary(
+	jsUrl := out[len("Live reload url: "):]
+	t.Logf("Livereload URL: '%s'", jsUrl)
+
+	validUrl := regexp.MustCompile("^http\\:\\/\\/localhost\\:[0-9]+\\/livereload\\.js\\?snipver\\=1$")
```
👍

gregmagolan commented on e2e/profiler/profiler_test.go:
```diff
@@ -0,0 +1,164 @@
+	jsUrl := out[len("Profiler url: "):]
+	t.Logf("Profiler URL: '%s'", jsUrl)
+
+	validUrl := regexp.MustCompile("^http\\:\\/\\/localhost\\:[0-9]+\\/profiler\\.js$")
```
👍

gregmagolan commented on e2e/live_reload/live_reload_test.go:
```diff
@@ -77,11 +80,39 @@ sh_binary(
+	}
+
+	bodyString := string(body)
+	validBodyStart := regexp.MustCompile("^\\(function e\\(t\\,n\\,r\\)")
```
👍

achew22 commented:
@gregmagolan, I'm not seeing your latest round. Did you maybe push it to a different branch?

gregmagolan commented:
@achew22 On its way. I'm working late tonight

gregmagolan commented:
@achew22 Changes are in. Thanks for all of the great feedback. Appreciate the detail you went into. Ran clang-format on profile.js as well.

achew22 commented on ibazel/profiler/profiler.go:
```diff
@@ -0,0 +1,429 @@
+	i.iterationBuildStart = true
+}
+
+func jsHandler(i *Profiler) http.HandlerFunc {
```
Methods attached to objects hold their reference to the receiver. All the handlers you're writing can be attached to the (i *Profiler) without having to take it as a parameter. That means you don't need to nest the function in a function and you can have jsHandler directly implement the interface `func(ResponseWriter, *Request)`

I put together a partial demo that I hope makes this a little bit clearer https://play.golang.org/p/bVnlREy50U6

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,103 @@
+          }
+          return "";
+        })
+        .reduce(reducer);
```
I think you should inline the reducer in here

achew22 commented on ibazel/profiler/profiler.js:
```diff
@@ -0,0 +1,103 @@
+        accumulator, currentValue) { return accumulator || currentValue; };
+    return [...window.document.getElementsByTagName('script') ]
+        .map(function(element) {
+          if (element.src) {
```
You're mixing filtering and mapping logic in here. You could also do

```
    return [...window.document.getElementsByTagName('script') ]
        .filter(function(e) { return element.src})
￼        .map(function(e) {
           const parsed = document.createElement('a');
           parsed.href = e.src;
           return parsed;
         })
        .filter(function(e) { return e.src})
        .map(function(e) {
           return `${parsed.protocol}//${parsed.host}/profiler-event`;
         })
        .reduce(function(acc, u) {...})
```

and then not have that if statement

gregmagolan replied:
Much cleaner. Thanks.

gregmagolan commented on e2e/live_reload/live_reload_test.go:
```diff
@@ -77,11 +79,41 @@ sh_binary(
+	}
+
+	expectedEnd := "},{}]},{},[8]);"
+	actualEnd := bodyString[len(bodyString)-len(expectedEnd):len(bodyString)]
```
Nice

gregmagolan commented on ibazel/profiler/profiler.go:
```diff
@@ -0,0 +1,429 @@
+	i.server.Addr = makeAddr(port)
+
+	// Handle profiler.js requests
+	router.HandleFunc("/profiler.js", jsHandler(i))
```
Cool. Didn't know you could do that.

achew22 commented:
Test this please

gregmagolan commented:
@achew22 Ready to go. Thanks!

achew22 commented:
Test this please

achew22 commented:
Test this please

gregmagolan commented:
@googlebot Confirm that this author is ok with their commits being contributed to this project

achew22 commented:
For @gregmagolan 

> Yes, you can accept code from this contributor.
> We found the following CLAs for your query.

For @achew22

> Yes, you can accept code from this contributor.
> We found the following CLAs for your query.

achew22 commented:
Good work, Greg. This is some really well done code. I appreciate your patience and really look forward to working more with you in the future.

alexeagle commented:
Yay nice job Greg! :tada:
